File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {number[]} preorder
     * @param {number[]} inorder
     * @return {TreeNode}
     */
    buildTree(preorder, inorder) {
        if (preorder.length < 1 || inorder.length < 1) return null

        const root = new TreeNode(preorder[0])
        const midIndex = inorder.indexOf(preorder[0])
        root.left = this.buildTree(
            preorder.slice(1, midIndex + 1), // midIndex here is used to determine how many nodes are in the left subtree
            inorder.slice(0, midIndex)
        )
        root.right = this.buildTree(
            preorder.slice(midIndex + 1),
            inorder.slice(midIndex + 1)
        )
        return root
    }
}

File: Data Structures & Algorithms/invert-a-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        def dfs(n):
            if not n:
                return

            temp = n.left
            n.left = n.right
            n.right = temp

            dfs(n.left)
            dfs(n.right)

        dfs(root)
        return root